core/src/main/java/org/apache/calcite/sql/type/ReturnTypes.java

```diff
@@ -613,6 +613,59 @@ public static SqlCall stripSeparator(SqlCall call) {
   public static final SqlReturnTypeInference ARG0_EXCEPT_INTEGER_NULLABLE =
       ARG0_EXCEPT_INTEGER.andThen(SqlTypeTransforms.TO_NULLABLE);
 
+  /**
+   * Returns the type of the first non NULL argument or INTEGER if all arguments
+   * are NULL.
+   */
+  public static final SqlReturnTypeInference LARGEST_INT_OR_FIRST_NON_NULL_DEFAULT_INTEGER =
+      opBinding -> {
+        final RelDataTypeFactory typeFactory = opBinding.getTypeFactory();
+        RelDataType integerReturnType = null;
+        boolean allArgsInteger = true;
+        for (RelDataType opType : opBinding.collectOperandTypes()) {
+          if (SqlTypeName.INT_TYPES.contains(opType.getSqlTypeName())) {
+            if (integerReturnType == null) {
+              integerReturnType = opType;
+            } else if (isIntegerTypeLarger(integerReturnType, opType)) {
+              integerReturnType = opType;
+            }
+          } else {
+            allArgsInteger = false;
+            break;
+          }
+        }
+        if (integerReturnType != null && allArgsInteger) {
+          return typeFactory.createTypeWithNullability(integerReturnType, true);
+        }
+        for (RelDataType opType : opBinding.collectOperandTypes()) {
+          if (!opType.isNullable()) {
+            return typeFactory.createTypeWithNullability(opType, true);
+          }
+        }
+        return typeFactory.createTypeWithNullability(
```

Comment:
When can this happen?

Reply:
For line 642 above, it would get there when at least one argument is a non integer type and not null. This is probably not quite correct. I only want to exclude the `NULL` type. A nullable value is fine as long as it is an integer or binary type.

I'll see about correcting this.

Reply:
I have reworked the logic here.

If all arguments are of the NULL type, then the return type is INTEGER nullable.

If all arguments are integer types, then the return type is the largest type found. It is nullable if at least one argument is nullable.

As a fallback, the first argument that is not of the NULL type is the return type. It is nullable if at least one argument is nullable.

Reply:
The fallback could be coded more explicitly as a check for the NULL type and then an assertion failure.
I like to be as explicit as possible, it's more future-proof.

Reply:
In fact, if any argument has the NULL type the result type can also have the NULL type, since there is only one value with this type, NULL.

Reply:
I agree with you in theory, but I don't know how to do this in practice. When I make the change so that the NULL type is returned if any argument is of the NULL type, then the tests fail. They fail with this error occurring when trying to convert the result.
```
java.lang.NoSuchMethodException: java.lang.Void.valueOf(void)
```

It looks like it is trying to the result to a Void type in Java. It never went to an implementation of the `BITAND()` function. I also tried use a return type of `java.lang.Long` for the relevant implementation of the `BITAND()` function. This gives the a similar error but for the Long type error:
```
java.lang.NoSuchMethodException: java.lang.Long.voidValue()
```

Another way to look at this is that `BITAND(1, NULL)` should return a `NULL` value of the type of the first argument. I am treating `BITAND(1, NULL)` as the same as `BITAND(NULL, 1)`.

Reply:
I was hoping that the nullpolicy would take care of that, but I guess I am wrong (the STRONG policy ensures that a function with null arguments is never called). But I guess you still need a return type, it can't be NULL. So probably the right solution is to return a compile-time error when all arguments have NULL type, and otherwise the type of the first non-NULL typed argument will do fine if any argument has type NULL.

Reply:
Updated. Now it gives an error when all arguments are NULL.